File: structy/structy-js/Binary-Tree/treeLevels.js
// Write a function, treeLevels, that takes in the root of a binary tree. The function should return a 2-Dimensional array where each subarray represents a level of the tree.

class Node {
	constructor(val) {
		this.val = val;
		this.left = null;
		this.right = null;
	}
}

//! iterative
const treeLevels = (root) => {
	if (root === null) return [];

	const levels = [];
	const queue = [{ node: root, levelNum: 0 }];
	while (queue.length > 0) {
		const { node, levelNum } = queue.shift();

		if (levels.length === levelNum) {
			levels[levelNum] = [node.val];
		} else {
			levels[levelNum].push(node.val);
		}

		if (node.left !== null)
			queue.push({ node: node.left, levelNum: levelNum + 1 });
		if (node.right !== null)
			queue.push({ node: node.right, levelNum: levelNum + 1 });
	}

	return levels;
};

//! recursive
// const treeLevels = (root) => {
// 	const levels = [];
// 	_treeLevels(root, levels, 0);
// 	return levels;
// };

// const _treeLevels = (root, levels, levelNum) => {
// 	if (root === null) return;

// 	if (levels.length === levelNum) {
// 		levels[levelNum] = [root.val];
// 	} else {
// 		levels[levelNum].push(root.val);
// 	}

// 	_treeLevels(root.left, levels, levelNum + 1);
// 	_treeLevels(root.right, levels, levelNum + 1);
// };

//! TEST_00
const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

console.log(treeLevels(a)); // ->
// [
//   ['a'],
//   ['b', 'c'],
//   ['d', 'e', 'f']
// ]

//! TEST_01
// const a = new Node("a");
// const b = new Node("b");
// const c = new Node("c");
// const d = new Node("d");
// const e = new Node("e");
// const f = new Node("f");
// const g = new Node("g");
// const h = new Node("h");
// const i = new Node("i");

// a.left = b;
// a.right = c;
// b.left = d;
// b.right = e;
// c.right = f;
// e.left = g;
// e.right = h;
// f.left = i;

// //         a
// //      /    \
// //     b      c
// //   /  \      \
// //  d    e      f
// //      / \    /
// //     g  h   i

// treeLevels(a); // ->
// // [
// //   ['a'],
// //   ['b', 'c'],
// //   ['d', 'e', 'f'],
// //   ['g', 'h', 'i']
// // ]

//! TEST_02
// const q = new Node("q");
// const r = new Node("r");
// const s = new Node("s");
// const t = new Node("t");
// const u = new Node("u");
// const v = new Node("v");

// q.left = r;
// q.right = s;
// r.right = t;
// t.left = u;
// u.right = v;

// //      q
// //    /   \
// //   r     s
// //    \
// //     t
// //    /
// //   u
// //  /
// // v

// treeLevels(q); //->
// // [
// //   ['q'],
// //   ['r', 's'],
// //   ['t'],
// //   ['u'],
// //   ['v']
// // ]

//! TEST_03
// treeLevels(null); // -> []

File: structy/structy-py/3-Binary-Tree/treeLevels.py
# Write a function, tree_levels, that takes in the root of a binary tree. The function should return a 2-Dimensional list where each sublist represents a level of the tree.

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

#! iterative


def tree_levels(root):
    if root is None:
        return []
    stack = [(root, 0)]  # adding a tuple or tracking of a level
    levels = []
    while stack:
        node, level_num = stack.pop()  # destructure the stack because now there's 2 elements
        if len(levels) == level_num:  # this checks which level and creates a subarray
            levels.append([node.val])
        else:
            # if level exsists then append node value to the level array
            levels[level_num].append(node.val)
        if node.right:
            stack.append((node.right, level_num+1))
        if node.left:
            stack.append((node.left, level_num+1))
    return levels

#! recursion
# def tree_levels(root):
#     if root is None:
#         return []


#!TEST_00
a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

print(tree_levels(a))  # ->
# [
#   ['a'],
#   ['b', 'c'],
#   ['d', 'e', 'f']
# ]

#!TEST_01
a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')
g = Node('g')
h = Node('h')
i = Node('i')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f
e.left = g
e.right = h
f.left = i

#         a
#      /    \
#     b      c
#   /  \      \
#  d    e      f
#      / \    /
#     g  h   i

print(tree_levels(a))  # ->
# [
#   ['a'],
#   ['b', 'c'],
#   ['d', 'e', 'f'],
#   ['g', 'h', 'i']
# ]

#!TEST_02
q = Node('q')
r = Node('r')
s = Node('s')
t = Node('t')
u = Node('u')
v = Node('v')

q.left = r
q.right = s
r.right = t
t.left = u
u.right = v

#      q
#    /   \
#   r     s
#    \
#     t
#    /
#   u
#  /
# v

print(tree_levels(q))  # ->
# [
#   ['q'],
#   ['r', 's'],
#   ['t'],
#   ['u'],
#   ['v']
# ]

#!TEST_03
print(tree_levels(None))  # -> []
